lib/psych/comments/emitter.rb

```diff
@@ -237,7 +240,7 @@ def emit_lookahead_comments(node)
       end
 
       def emit_comment(comment)
-        unless /\A#[^\r\n]*\z/.match?(comment)
+        unless /\A\s*#[^\r\n]*\z/.match?(comment)
```

Comment:
Could you throw away whitespace details before calling `emit_comment`? The library is meant to preserve comments, but not the spacing details.

Reply:
We certainly could!

But I don't think we should.

It would wreck `psych-comments`' ability to round-trip YAML like this (with justified line-end comments):
```yaml
fruit:
- fig      # ripens in late spring
- apricot  # ripens in the summer
- apple    # ripens in the fall
```

It wouldn't be able to faithfully render the YAML it parses. I'd think that fidelity would be paramount, wouldn't you?

Reply:
The library is meant to preserve comments, but not the spacing details.

If you want justification, you may implement a justification algorithm and I'm willing to accept it if its quality is good enough. Major design decisions you will need to consider implementing it are:

- How is the width calculated? What happens if there is ambiwidth or long grapheme cluster?
- What constitutes a single "run" of comments that should be aligned at the same position? Should it be limited to the same level of leaves? What happens if there are lines with comments and those without?

Reply:
Hey Masaki!

Oh, I definitely wouldn't want to introduce an algorithm for justification — I wouldn't expect this library to have an opinion about whitespace! I wouldn't want it to _introduce_ justification where it didn't exist before. (I didn't see it like Rubocop or prettier.js or anything that would _reformat_ YAML.)

My understanding is that psych-comments is trying to allow you to manipulate YAML more losslessly (from the README: "This gem allows you to manipulate YAML, preserving comment information.")

My use-case is manipulating thousands of YAML files written by as many engineers. They (my internal customers) have different conventions for spacing and commenting these configuration files.

What are the cons to preserving whatever whitespace exists between Scalar nodes and line-end comments?
What would be the advantage of discarding it?

Reply:
> I didn't see it like Rubocop or prettier.js or anything that would reformat YAML.

It **does** behave like a reformatter to a certain degree -- consider the following example:

```yaml
-           foo: \"\\x62\\x61\\x72\"
```

When you `Psych::Comments.emit_yaml(Psych::Comments.parse_stream(source))` it, this reformats to:

```yaml
- foo: "bar"
```

The behavior is meant to imitate how Psych (or libyaml) behaves.

> My understanding is that psych-comments is trying to allow you to manipulate YAML more losslessly

No, this is as it reads -- it preserves comment information.

> What are the cons to preserving whatever whitespace exists between Scalar nodes and line-end comments?

Missing a clear boundary on the library's scope sometimes puts a heavy burden on the maintainer, and _I_ cannot take the risk in this case. As you say, many YAML authors have many opinions on how they write the value, the prominent example of which is

```yaml
foo:
- bar
- baz
```

vs.

```yaml
foo:
  - bar
  - baz
```

or whatever details. As I'm not prepared to accept all these requirements, I'm going to decide following principles:

- The library is meant to preserve comments, but not the spacing details.
- Or, to put it differently: this is Psych (libyaml) augmented with comment node. Other syntactic details are thrown away.

If you are confident you can handle (i.e. accept or reject) all those sorts of additional demands and keep the library clean enough to maintain, I won't stop you from forking this library (nor have the authority to do so, of course). Though the gem named `psych-comments` will remain simple enough to handle just comments.